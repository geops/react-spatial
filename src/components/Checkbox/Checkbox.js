import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  /**
   * Function triggered on button's click event.
   */
  onClick: PropTypes.func.isRequired,

  /**
   * Boolean attribute if input is checked or not.
   */
  checked: PropTypes.bool,

  /**
   * CSS class of the checkbox.
   */
  className: PropTypes.string,

  /**
   * Is either checkbox or radio.
   */
  inputType: PropTypes.string,

  /**
   * HTML tabIndex attribute.
   */
  tabIndex: PropTypes.number,
};

const defaultProps = {
  className: 'tm-check',
  inputType: 'checkbox',
  checked: false,
  tabIndex: 0,
};

class Checkbox extends PureComponent {
  render() {
    const { className, onClick, tabIndex, inputType, checked } = this.props;
    /*
      eslint-disable jsx-a11y/label-has-associated-control,
      jsx-a11y/label-has-for
    */
    return (
      <label
        className={`${className} tm-${inputType}`}
        tabIndex={tabIndex}
        onKeyPress={e => {
          if (e.which === 13) {
            onClick();
          }
        }}
      >
        <input
          type={inputType}
          tabIndex={-1}
          checked={checked}
          onChange={() => {}}
          onClick={() => onClick()}
        />
        <span />
      </label>
    );
    /*
      eslint-enable jsx-a11y/label-has-associated-control,
      jsx-a11y/label-has-for
    */
  }
}

Checkbox.propTypes = propTypes;
Checkbox.defaultProps = defaultProps;

export default Checkbox;
