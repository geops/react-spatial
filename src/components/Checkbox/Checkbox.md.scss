.tm-checkbox-example {
  margin: 5px;
}
