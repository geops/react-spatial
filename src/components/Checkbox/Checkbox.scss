.tm-checkbox {
  span {
    &::after {
      left: 5px;
      top: 1px;
      width: 5px;
      height: 10px;
      border: solid $brand-primary;
      border-width: 0 1px 1px 0;
      -webkit-transform: rotate(45deg);
      -ms-transform: rotate(45deg);
      transform: rotate(45deg);
    }
  }
}

.tm-radio {
  span {
    border-radius: 50%;

    &::after {
      top: 2px;
      left: 2px;
      width: 12px;
      height: 12px;
      border-radius: 50%;
      background: $brand-primary;
    }
  }
}

.tm-checkbox,
.tm-radio {
  display: inline-block;
  position: relative;
  cursor: pointer;
  height: 18px;
  width: 18px;

  span {
    background-color: white;
    border: 1px solid lightgray;
    width: calc(100% - 2px);
    height: calc(100% - 2px);
    position: absolute;
    top: 0;
    left: 0;
    margin: 0;

    &::after {
      content: '';
      position: absolute;
      display: none;
    }
  }

  &:hover input ~ span {
    background-color: #ccc;
  }

  input {
    position: absolute;
    top: 0;
    left: 0;
    margin: 0;
    height: 18px;
    width: 18px;
    display: none;
  }

  input:checked ~ span::after {
    display: block;
  }
}
