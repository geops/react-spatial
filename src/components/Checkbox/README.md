#

This demonstrates the use of Checkbox.

```jsx
import React from  'react';
import Checkbox from 'react-spatial/components/Checkbox';

class CheckboxExample extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      check: true,
      radio1: false,
      radio2: true,
    }
  }

  onRadioClick() {
    const { radio1, radio2 } = this.state;
    this.setState({
      radio1: !radio1,
      radio2: !radio2,
    });
  }

  onCheckboxClick() {
    const { check } = this.state;
    this.setState({ check: !check });
  }

  render() {
    const { check, radio1, radio2 } = this.state;
    return (
      <div>
        <div>
          <Checkbox
            className="tm-checkbox-example"
            checked={check}
            onClick={() => this.onCheckboxClick()}
          />
        </div>
        <div>
          <Checkbox
            className="tm-checkbox-example"
            checked={radio1}
            inputType="radio"
            onClick={() => this.onRadioClick()}
          />
          <Checkbox
            className="tm-checkbox-example"
            checked={radio2}
            inputType="radio"
            onClick={() => this.onRadioClick()}
          />
        </div>
      </div>
    )
  }
}

<CheckboxExample />;
```
