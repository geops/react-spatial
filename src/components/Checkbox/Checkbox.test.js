import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Checkbox from './Checkbox';

configure({ adapter: new Adapter() });

describe('Checkbox', () => {
  describe('matches snapshots', () => {
    test('for inputType Checkbox.', () => {
      const component = renderer.create(
        <Checkbox className="tm-checkbox-chbx-example" onClick={() => {}} />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('for inputType Radio.', () => {
      const checked = true;
      const component = renderer.create(
        <Checkbox
          className="tm-checkbox-radio-ex"
          inputType="radio"
          checked={checked}
          onClick={() => {}}
        />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('triggers onClick', () => {
    let wrapper;
    let spy;
    let checked = true;

    const funcs = {
      onCheckboxClick: () => {
        checked = !checked;
      },
    };

    const expectCalled = () => {
      expect(spy).toHaveBeenCalledTimes(1);
    };

    beforeEach(() => {
      wrapper = shallow(
        <Checkbox
          className="tm-checkbox-ex"
          checked={checked}
          onClick={() => funcs.onCheckboxClick()}
        />,
      );
      spy = jest.spyOn(funcs, 'onCheckboxClick');
    });

    afterEach(() => {
      spy.mockRestore();
    });

    test('Checkbox should be clicked.', () => {
      wrapper.find({ type: 'checkbox' }).simulate('click');
      expectCalled();
    });

    test('Checkbox should be clicked.', () => {
      wrapper
        .find('label')
        .first()
        .simulate('keypress', { which: 13 });
      expectCalled();
    });
  });
});
