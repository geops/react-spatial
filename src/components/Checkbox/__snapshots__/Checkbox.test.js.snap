// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox matches snapshots for inputType Checkbox. 1`] = `
<label
  className="tm-checkbox-chbx-example tm-checkbox"
  onKeyPress={[Function]}
  tabIndex={0}
>
  <input
    checked={false}
    onChange={[Function]}
    onClick={[Function]}
    tabIndex={-1}
    type="checkbox"
  />
  <span />
</label>
`;

exports[`Checkbox matches snapshots for inputType Radio. 1`] = `
<label
  className="tm-checkbox-radio-ex tm-radio"
  onKeyPress={[Function]}
  tabIndex={0}
>
  <input
    checked={true}
    onChange={[Function]}
    onClick={[Function]}
    tabIndex={-1}
    type="radio"
  />
  <span />
</label>
`;
